rust/actyx/ax-aql/src/language/parser.rs

```diff
@@ -721,6 +722,100 @@ pub(crate) fn query_from_str(s: &str) -> Result<Query<'_>> {
+    pub(crate) error_type: QueryWorkflowAnalysisErrorType,
+}
+
+pub(crate) enum QueryWorkflowAnalysisErrorType {
+    UndeclaredEvent,
+    UndeclaredWorkflowCall,
+}
+
+pub(crate) struct QueryWorkflowAnalysis<'a> {
+    pub(crate) query: &'a Query<'a>,
+}
+
+impl<'a> QueryWorkflowAnalysis<'a> {
+    pub fn from(query: &'a Query) -> Self {
+        Self { query }
+    }
+
+    pub fn check(&'a self) -> Vec<QueryWorkflowAnalysisError> {
+        self.query
+            .workflows
+            .iter()
+            .flat_map(|(_ident, workflow)| self.check_workflow(workflow))
+            .collect()
+    }
+
+    fn check_workflow(&'a self, workflow: &'a Workflow) -> Vec<QueryWorkflowAnalysisError> {
+        self.check_steps(&workflow.steps)
+    }
+
+    fn check_steps(&'a self, steps: &'a NonEmptyVec<WorkflowStep<'a>>) -> Vec<QueryWorkflowAnalysisError> {
+        steps.iter().flat_map(|step| self.check_step(step)).collect()
+    }
+
+    fn check_step(&'a self, step: &'a WorkflowStep<'a>) -> Vec<QueryWorkflowAnalysisError> {
+        use super::workflow::WorkflowStep::*;
+
+        match step {
+            Event { label, binders: _, .. } => {
+                let mut errors = Vec::new();
+                if !self.query.events.contains_key(&label) {
+                    errors.push(QueryWorkflowAnalysisError {
+                        error_type: QueryWorkflowAnalysisErrorType::UndeclaredEvent,
+                        span: label.clone(),
+                    });
+                }
+                // this part will also handle additional stuffs like type matching in `binders`
+                errors
+            }
+            Retry { steps } => self.check_steps(steps),
+            Timeout { steps, .. } => self.check_steps(steps),
+            Parallel { cases, .. } => cases.iter().flat_map(|case| self.check_steps(case)).collect(),
+            Call {
+                workflow: workflow_ident,
+                cases,
+                ..
+            } => {
+                let mut errors = Vec::new();
+                if let None = self.query.workflows.get(&*workflow_ident) {
+                    errors.push(QueryWorkflowAnalysisError {
+                        error_type: QueryWorkflowAnalysisErrorType::UndeclaredWorkflowCall,
+                        span: workflow_ident.clone(),
+                    });
+                }
+
+                errors.extend(cases.iter().flat_map(|(_, steps)| self.check_steps(steps)));
+
+                errors
+            }
+            Compensate { body, with } => {
+                let mut errors = Vec::new();
+                errors.extend(self.check_steps(body));
+                errors.extend(self.check_steps(with));
+                errors
+            }
+            Choice { cases } => cases.iter().flat_map(|steps| self.check_steps(steps)).collect(),
+        }
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct WorkflowTracker<'a>(RefCell<BTreeSet<&'a Workflow<'a>>>);
```

Comment:
I'm curious as to why a `RefCell`, if this is an analysis pass, this should simply perform the analysis, "lock it" and move on. 

What am I missing?

Reply:
Never mind this. It was used to remember which workflow had been analyzed. But now it is not used again.
The reason for the RefCell itself is so that a non-mutable reference can be passed recursively to eliminate headache from the possibility of multiple `mut` references